modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +327,66 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private class RemoteHostValidator {
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : line.split(" ")) {
+                if (linePart.startsWith("from=")) {
+                    for (String pattern : strip(linePart).split(",")){
+                        boolean patternIsNegated = pattern.startsWith("!");
+                        boolean patternMatchesHost = patternMatchesHost(pattern, session);
+                        if (patternIsNegated==patternMatchesHost){
```

Comment:
The result can short-cut only if the pattern is a deny-pattern.  If an allow-pattern matches then subsequent deny-patterns should be considered.

An allow-pattern match should be remembered, as it should affect the return value.

Also coding style: space either side of `==` and  between parenthesis and brace: `) {`